### ISSUE TITLE
ground truth width and height policy

### ISSUE THREAD

**Issue by developer0hye** [yjh0410/pytorch-yolo-v2#7]
Hi, yjh!

https://github.com/yjh0410/pytorch-yolo-v2/blob/baecdae27e4b75883088729dbd60a8f2a1270c80/tools.py#L317-L318

Why do you divide width and height of object by largest stride size?

I also saw this policy in other yolov3 implementations, but I couldn't understand that reason.
Can you expect that we can predict the normalized width and height with normalized anchor box?

**Comment by yjh0410** [comment]
Because my anchor boxes used in YOLO-v2 are all divided by 32(the largest stride).

**Comment by developer0hye** [comment]
Thank you for your answer!
So... I think that the division can suppress the range of width and height and it will make numerically stable model. Is my opinion right?

If I use the not divided anchor box, can I omit this division operation?

**Comment by yjh0410** [comment]
Yeah, I think you can omit the division operation without division.

My model make predictions on the scale of feature maps not input images, and I think this could be helpful for training stably. But, I didn't try without division ... maybe there will be some difference.

**Comment by developer0hye** [comment]
Thanks!

And... I have one more question.

https://github.com/yjh0410/pytorch-yolo-v2/blob/master/models/yolo_v3.py#L22-L25

Why

**Comment by yjh0410** [comment]
Each feature map has different stride. My all anchor boxes in YOLO-v3 are divided by 32(the largest stride) while I use three feature map with 8, 16, 32 strides. So, multiply 4（32/8=4） is corresponding to stride=8 feature map，multiply 2（32/16=4） is corresponding to stride=16 feature map, and multiply 1（32/32=1） is corresponding to stride=32 feature map. 

If you don't apply division operation, just omit it.

**Comment by developer0hye** [comment]
@yjh0410 
Thanks, I was always curious about this because I couldn't find the explanation related to this part.

**Comment by yjh0410** [comment]
You're welcome ！Fighting！Fighting！

**Comment by gohurali** [comment]
Hi yjh, continuing thread as I have questions on a similar topic. 

Great project by the way, I am using your project as a method of further understanding the YOLO papers. 

The anchor box strides which you have chosen: `[8, 16, 32]` are, I am assuming, the distance between each anchor box placement in the grid cell? Does this mean that the number of grid cells is equivalent to the number of strides?

**Comment by yjh0410** [comment]
Hi !
In official YOLOv3, it takes advantage of three feature maps whose strides are 8, 16, 32 from backbone network. But there is no obvious relation among strides[8, 16, 32], the number of anchor boxes and grid cells.


As for the size of gride, it is equal to "input size // stride". Specifically, if input size is 416x416, and stride is 8, then the size of grid is (416 // 8)x(416 // 8)=52x52. And in each grid cell, there are N(default=3) anchor boxes.
So the smaller stride, the denser the grid.


therefore, for example, if the input size is 416, for each stride:
input size=416
stride=8: grid-&gt;52x52;(Prefer to detect small objects)
stride=16: grid-&gt;26x26(Prefer to detect medium objects);
stride=32:grid -&gt;13x13(Prefer to detect large objects)


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "yjh0410/yolov2-yolov3_PyTorch"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月21日(星期三) 上午6:52
收件人:&nbsp;"yjh0410/yolov2-yolov3_PyTorch"<yolov2-yolov3_PyTorch@noreply.github.com&gt;;
抄送:&nbsp;"ら ．Secret"<1394571815@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [yjh0410/yolov2-yolov3_PyTorch] ground truth width and height policy (#7)





 
Hi yjh, continuing thread as I have questions on a similar topic.
 
Great project by the way, I am using your project as a method of further understanding the YOLO papers.
 
The anchor box strides which you have chosen: [8, 16, 32] are, I am assuming, the distance between each anchor box placement in the grid cell? Does this mean that the number of grid cells is equivalent to the number of strides?

**Comment by gohurali** [comment]
Hi yjh,
Thank you for your detailed response. 

I guess my confusion was coming from the fact that if you divide the input size by some number, then it would give you the size of the cell itself rather then the size of the entire grid. 

So to confirm my understanding, if the `input size = 416` and `stride = 8`, we get a 52x52 feature map. Then, we apply our N anchor boxes along each cell (total being 52 x 52 = 2704 cells)?

**Comment by yjh0410** [comment]
Yes !
After that, there are 52x52xN anchor boxes(with input size=416 and stride=8). 
So many anchor boxes...



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "yjh0410/yolov2-yolov3_PyTorch"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年10月21日(星期三) 下午4:43
收件人:&nbsp;"yjh0410/yolov2-yolov3_PyTorch"<yolov2-yolov3_PyTorch@noreply.github.com&gt;;
抄送:&nbsp;"ら ．Secret"<1394571815@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [yjh0410/yolov2-yolov3_PyTorch] ground truth width and height policy (#7)





 
Hi yjh,
 Thank you for your detailed response.
 
I guess my confusion was coming from the fact that if you divide the input size by some number, then it would give you the size of the cell itself rather then the size of the entire grid.
 
So to confirm my understanding, if the input size = 416 and stride = 8, we get a 52x52 feature map. Then, we apply our N anchor boxes along each cell (total being 52 x 52 = 2704 cells)?

**Comment by gohurali** [comment]
Your explanation is clear. Thank you so much for the swift replies. 
Yes, it seemed odd at first to have so many, but it makes sense now. If I have any further questions I will open a new thread or ask here depending on the question.
